src/resolve/resolved-dns-transaction.c

```diff
@@ -1124,6 +1125,33 @@ void dns_transaction_process_reply(DnsTransaction *t, DnsPacket *p, bool encrypt
 
                 if (!t->bypass &&
                     IN_SET(DNS_PACKET_RCODE(p), DNS_RCODE_FORMERR, DNS_RCODE_SERVFAIL, DNS_RCODE_NOTIMP)) {
+                        /* If the server has replied with detailed error data, using a degraded feature set
+                         * will likely not help anyone. Examine the detailed error to determine the best
+                         * course of action. */
+                        r = dns_packet_ede_rcode(p, &ede_rcode);
+                        if (r < 0)
+                                goto fail;
+                        if (r > 0 && DNS_PACKET_RCODE(p) == DNS_RCODE_SERVFAIL) {
+                                /* These codes are related to DNSSEC configuration errors. If accurate,
+                                 * this is the domain operator's problem, and retrying won't help. */
+                                if (ede_rcode_is_dnssec(ede_rcode)) {
+                                        dns_transaction_complete(t, DNS_TRANSACTION_DNSSEC_FAILED);
+                                        return;
+                                }
+
+                                /* These codes probably indicate a transient error. Let's try again. */
+                                if (IN_SET(ede_rcode, DNS_EDE_RCODE_NOT_READY, DNS_EDE_RCODE_NET_ERROR)) {
+                                        dns_transaction_retry(t, false);
+                                        return;
+                                }
+
+                                /* OK, the query failed, but we still shouldn't degrade the feature set for
+                                 * this server. */
+                                log_debug("Server returned error: %s (EDE %d)",
+                                                FORMAT_DNS_RCODE(DNS_PACKET_RCODE(p)),
+                                                ede_rcode);
```

Comment:
I think it would be fantastic if we'd have FORMAT_DNS_EDE_RCODE() or so? jus tmakes things easier to debug? any chance you can add that? it's realatively simple using the string table macros we have.

Reply:
Ok, I've just copied most of the error strings from bind9.